packages/firestore/src/local/local_store.ts

```diff
@@ -642,6 +618,187 @@ export class LocalStore {
       });
   }
 
+  /**
+   * Applies the documents from a bundle to the "ground-state" (remote)
+   * documents.
+   *
+   * LocalDocuments are re-calculated if there are remaining mutations in the
+   * queue.
+   */
+  applyBundleDocuments(documents: BundledDocuments): Promise<MaybeDocumentMap> {
+    let documentMap = maybeDocumentMap();
+    let versionMap = documentVersionMap();
+    for (const bundleDoc of documents) {
+      const documentKey = this.bundleConverter.toDocumentKey(
+        bundleDoc.metadata.name!
+      );
+      documentMap = documentMap.insert(
+        documentKey,
+        this.bundleConverter.toMaybeDocument(bundleDoc)
+      );
+      versionMap = versionMap.insert(
+        documentKey,
+        this.bundleConverter.toSnapshotVersion(bundleDoc.metadata.readTime!)
+      );
+    }
+
+    const documentBuffer = this.remoteDocuments.newChangeBuffer({
+      trackRemovals: true // Make sure document removals show up in `getNewDocumentChanges()`
+    });
+    return this.persistence.runTransaction(
+      'Apply bundle documents',
+      'readwrite-primary',
+      txn => {
+        return this.applyDocuments(
+          txn,
+          documentBuffer,
+          documentMap,
+          undefined,
+          versionMap
+        )
+          .next(changedDocs => {
+            documentBuffer.apply(txn);
+            return changedDocs;
+          })
+          .next(changedDocs => {
+            return this.localDocuments.getLocalViewOfDocuments(
+              txn,
+              changedDocs
+            );
+          });
+      }
+    );
+  }
+
+  /**
+   * Applies documents to remote document cache, returns the document changes
+   * resulting from applying those documents.
+   *
+   * @param updatedKeys Keys of the documents to be applied.
+   * @param documents Documents to be applied.
+   * @param remoteVersion The read time of the documents to be applied, it is
+   * a `SnapshotVersion` if all documents have the same read time, or a `DocumentVersionMap` if
+   * they have different read times.
+   */
+  private applyDocuments(
+    txn: PersistenceTransaction,
+    documentBuffer: RemoteDocumentChangeBuffer,
+    documents: MaybeDocumentMap,
+    globalVersion: SnapshotVersion | undefined,
+    documentVersions: DocumentVersionMap | undefined
```

Comment:
Looking at this more, would it be possible to use the readTime from the bundle as the `globalVersion` and drop `documentVersions`? AFAIK the only real guarantee that we have to make is that a changed document has a changed read time that is higher than the new document version. 

If that is not possible, I think we should explore adding the read time to MaybeDocument (it's already part of DbRemoteDocument, so we do have the data). That might be a relatively simple way to improve the code flow here.

Reply:
I think it would be possible to add a readTime to `MaybeDocument`. It looks like we would need to change a lot of our plumbing, so I am ok with leaving as is. We could add a TODO if this is something that we think would clean up our code in general.

Reply:
Done.